### PR TITLE
SONARJAVA-1088 Rule: Assertions should be complete

### PR DESCRIPTION
**Pull request by mpaladin**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    // skip variable assignments
+  }
+
+  @Override
+  public void visitReturnStatement(ReturnStatementTree tree) {
+    // skip return statements
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree mit) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java]: New MAJOR The Cyclomatic Complexity of this method "visitMethodInvocation" is 13 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    if (FEST_ASSERT_THAT.anyMatch(mit) && mit.arguments().size() == 1) {
+      if (chainedTo.isEmpty() || !(chainedTo.contains(ChainedMethods.FEST) || chainedTo.contains(ChainedMethods.ANY_BUT_NOT_FEST_EXCLUSIONS))) {
+        context.addIssue(mit, this, "Complete the assertion.");
+        return;
+      }
+    } else if (MOCKITO_VERIFY.matches(mit)) {
+      if (chainedTo.isEmpty()) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java]: New MAJOR Merge this if statement with the enclosing one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1066)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

File: java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    // Fest 1.X
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.fest.assertions.GenericAssert")).name(NameCriteria.any()).withNoParameterConstraint(),
+    // Fest 2.X
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.fest.assertions.api.AbstractAssert")).name(NameCriteria.any()).withNoParameterConstraint()
+  );
+
+  private static final MethodInvocationMatcherCollection FEST_EXCLUSIONS = MethodInvocationMatcherCollection.create(
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.anyType()).name("as").withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.anyType()).name("describedAs").withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.anyType()).name("overridingErrorMessage").withNoParameterConstraint()
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java]: Could you create helper static method to simplify reading of `MethodInvocationMatchers`?

File: java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  @Override
+  public void visitReturnStatement(ReturnStatementTree tree) {
+    // skip return statements
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree mit) {
+    if (maybeRaiseIssue(mit)) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java]: The name of the method is not very explicit. Something like `isInvalidAssert` or similar would be better.

File: java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+      return;
+    }
+    if (FEST_ASSERTIONS.anyMatch(mit) && !FEST_EXCLUSIONS.anyMatch(mit)) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java]: I never thought about using the `MethodInvocationMatchers` like that. But I like the idea!

File: java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    } else {
+      chainedTo.push(ChainedMethods.ANY);
+    }
+    super.visitMethodInvocation(mit);
+    chainedTo.pop();
+  }
+
+  private boolean maybeRaiseIssue(MethodInvocationTree mit) {
+    if (FEST_ASSERT_THAT.anyMatch(mit) && mit.arguments().size() == 1) {
+      if (chainedTo.isEmpty() || !(chainedTo.contains(ChainedMethods.FEST) || chainedTo.contains(ChainedMethods.ANY_BUT_NOT_FEST_EXCLUSIONS))) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java]: As far as I can see using your `enum` is a bit overkill at that point. `FEST` and `ANY_BUT_NOT_FEST_EXCLUSIONS` have the same effect. That can be simplified.

**Comment by m-g-sonar** [comment]
I would prefer to have the test file simplified. A lot of logic in it is actually not usefull for the check. Can you also mark compliant lines ?

**Comment by mpaladin** [comment]
I simplified the rule and the test in https://github.com/mpaladin/sonar-java/commit/f7b26e2d7d55ae8a81f102aea39a3851680a9975.
I didn't create a method to create MethodInvocationMatchers, IMHO for 3 occurrences it is more readable as it is.
I don't like to mark lines as Compliant, whatever is not `Noncompliant` it is implicitly compliant.

**Comment by m-g-sonar** [comment]
Regarding the interest of marking `compliant` lines, I think that it allows to differentiate lines which are especially correct accordingly to the rule (i.e. correct implementation of what is targeted by the rule) from the lines of code which are just present for coverage and/or code coherence.

Except that, LGTM!

**Pull request merged**